src/DotVVM.Framework/Hosting/EmbeddedMarkupFileLoader.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using DotVVM.Framework.Configuration;
+
+namespace DotVVM.Framework.Hosting
+{
+    public class EmbeddedMarkupFileLoader : IMarkupFileLoader
+    {
+        /// <summary>
+        /// Gets the markup file for the specified virtual path.
+        /// </summary>
+        public MarkupFile GetMarkup(DotvvmConfiguration configuration, string virtualPath)
+        {
+            if (!virtualPath.StartsWith("embedded://"))
+            {
+                return null;
+            }
+
+            string resourceName = virtualPath.Remove(0, 11);
```

Comment:
Just a tip: `11` is a magic constant and it is not very obvious what does it mean and you had to count the letters which is annoying and it's easy to make a mistake. So although it is longer I prefer to use `"embedded://".Length`, but you can just copy-paste the string instead of counting characters ;)